### DISCUSSION TITLE
Support Providing Extra arguments from parent resolver to child resolver

### DISCUSSION THREAD

**Discussion by abishek-rz** [mirumee/ariadne#707] [General]
Hi I create some common context objects when resolving object in the root resolver, since I need to make an api call and prepare this context I don't want to repeat this call for every child resolver

The issue is when using query aliases the context from the last api call overrides all the query aliases, I would like to handle this by creating a query id when I'm at the root object and passing this as a additional query_id argument to all resolvers using extensions but It does not recursively pass from the parent to child resolver, could you please support this

**Comment by rafalp** [comment]
Root resolver returns data for top level query resolvers. It doesn't produce context.

To avoid repeating API calls you are supposed to cache your data on the context, but "having to repeat api call in child resolver" sounds like there's problem elsewhere. Usually child resolver shouldn't need to repeat parent's API call. Can you show the code?

**Reply by abishek-rz** [reply]
Sorry that I wasn't more clear with the first message
Let me add a bit more context to what I'm doing

Basically we are chaning together a bunch of REST apis into a single GraphQL API
For this we need to keep track of some unique keys like for example the unique key which we got from project api
I store this in the context as {"response_contexts": {"project_key": "UNIQUE__123293434"}}
But if we query for multiple projects for example
```graphql
{
    project1: project(key: "UNIQUE__123293434") {
        name
        owner {
            location {
                country
            }
        }
    }
    project2: project(key: "UNIQUE__2333") {
        name
        owner {
            location {
                country
            }
        }
    }
}
```
Then the context variable would overwrite
I need to know which project is being used in one of the child resolvers let's say I want to know which project is being used when I'm resolving the country field
currently I have no way of knowing this

For this I thought the best approach would be to use an extension like this
```python
class QueryExtension(Extension):
    async def resolve(
            self, next_: Resolver, parent: Any,
            info: GraphQLResolveInfo, **kwargs: Any) -> Any:
        query_id = kwargs.get('queryId')
        context_preparer = None
        if parent is None:
            query_id = str(uuid4())
            if isinstance(info.return_type, GraphQLObjectType) and info.return_type.name == "ProjectAPIResponse":
                context_preparer = ProjectContextPreparers
        kwargs['queryId'] = query_id
        result = await super().resolve(next_, parent, info, **kwargs)
        if isinstance(result, dict) and query_id:
            if context_preparer:
                info.context[query_id] = {'queryId': query_id}
                info.context[query_id]['response_contexts'] = await context_preparer.all(result)
        return result
```
I could refer this unique query id and fetch the specific information from any context eventhough all resolvers use a common context
but this query id kwarg is not passed to other resolvers

**Reply by rafalp** [reply]
> currently I have no way of knowing this

This is incorrect. Every resolver in GraphQL receives parent resolver's return value as first argument. So simpler approach is to include project in data returned from `owner` resolver:

```python
def resolve_project(...):
    return api.get(project_key)


def resolve_project_owner(project, info):
    owner = get_owner_from_somewhere(project.owner_id)
    # Option A:
    owner.project_context = project
    return owner
    # Option B:
    return ProjectOwner(owner=owner, project=project)
```

**Reply by abishek-rz** [reply]
So basically you are saying to include project in owner schema

Is there any other approach other than including the project owner schema, incase there are multiple nested fields (lets say another GraphQLObject Type inside LocationInfo type also...)  do I need to pass it to all the nested fields

```graphql
type ProjectOwner {
    name: String
    project_key: String
    location: LocationInfo
}

type LocationInfo {
    country: String
    project_key: String
}

type ProjectContext {
    project_key: String
    name: String
    owner: ProjectOwner
}
```

**Reply by rafalp** [reply]
> So basically you are saying to include project in owner schema

This is not what I am saying. ;) You don't have to represent python data structures your resolver returns in GraphQL schema. GraphQL only cares that resolvers for scalar fields (eg. `name: String`) will return value that can be serialized as such type in JSON.

For example:

```graphql
type User {
  fullname: String
}
```

```python
# this is example user resolver that returns user object with name and last name, but not full name
def resolve_user(...):
    return User(name="John", last_name="Doe")


# this resolver is for `fullname: String` field and takes `User` object from above as first arg:
def resolve_user_fullname(obj, ...):
    return f"{obj.name} {obj.last_name}"
```

Notice how `User` object returned by user resolver is completely different from `User` type in GraphQL Schema. Its user's field resolvers that know how to pull data from this python object.

**Reply by abishek-rz** [reply]
Thank you for this @rafalp This helps a lot
One more small question, Do I have to define the user class seperately or could I import it from the schema with

```python
from ariadne import ObjectType
user = ObjectType("User")
```

**Reply by rafalp** [reply]
You'll have to use `ObjectType` to tell ariadne that `resolve_user_fullname` is for `User`'s `fullname` field:

```python
from ariadne import ObjectType

user = ObjectType("User")


@user.field("fullname")
def resolve_user_fullname(obj, _):
    return f"{obj.name} {obj.last_name}"
```

**Reply by abishek-rz** [reply]
I mean when declaring this
```python
# this is example user resolver that returns user object with name and last name, but not full name
def resolve_user(...):
    return User(name="John", last_name="Doe")
```

**Reply by rafalp** [reply]
Ah. `User` class here is an example. It's up to you to decide how models are represented in your python logic. Everything is fine as long as it works for you.

**Reply by abishek-rz** [reply]
Thank you @rafalp
